### PR TITLE
Ru training pipeline

### PR DESCRIPTION
**Pull request by rfl-urbaniak**
add a layer of files needed for the training pipeline and the user-facing deployment using CausalInsight.

Most of the files changed are just guides and parameters. The first review is due by @azane, and we went over the main moves in a conversation.

### CODE REVIEW

File: cities/queries/causal_insight.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+            self.intervention_impact_mean + self.observed_outcomes.iloc[1:]
+        ).tolist()
+        predicted_low = [self.observed_outcomes.iloc[0]] + (
+            self.intervention_impact_low + self.observed_outcomes.iloc[1:]
+        ).tolist()
+        predicted_high = [self.observed_outcomes.iloc[0]] + (
+            self.intervention_impact_high + self.observed_outcomes.iloc[1:]
+        ).tolist()
+
+        self.predictions = pd.DataFrame(
```
Comment (azane) [inline cities/queries/causal_insight.py:200]: Went through the algebra just to confirm my intuition that this is indeed just the counterfactual — checks out to me:

obs = bias + f(X, Y) + tau * T + eps
eps = obs - (bias + f(X, Y) + tau * T)
int = bias + f(X, Y) + tau * T' + eps

int = bias + f(X, Y) + tau * T’ + obs - (bias + f(X, Y) + tau * T)
= tau * T’ + obs - tau * T
= tau * ( T’ - T) + obs

Comment (azane) [reply cities/queries/causal_insight.py:200]: A good sanity check here might be do run the full interventional predictive (not counterfactual) and make sure that the counterfactual computed here sits within the interventional predictive (where noise isn't assumed to be shared between the two regimes).

File: cities/queries/causal_insight.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        param_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_params.pth"
+        )
+
+        pyro.get_param_store().load(param_path)
+
+        self.forward_shift = forward_shift
+
+    def generate_samples(self):
+        self.data = prep_wide_data_for_inference(
```
Comment (azane) [inline cities/queries/causal_insight.py:50]: So a lot of these attributes are defined outside of the init, which can get a little bit funky — "standard" practice is to define all your attributes in the __init__ (setting yet-undefined things like so: `self.value: Optional[Blabla] = None`).

TLDR for rationale is just that it's nice to have all the attributes of a class listed out in one place — is as much for documentation as much as standardizing your error checking to be checking for `None` and not catching attribute errors.

Btw: it's probably worth throwing some errors if one of these methods you expect to have been called hasn't yet been called. That way you can clarify directly that x method should have been called, and future you won't be wondering why you're getting a random attribute error ;)

Comment (azane) [reply cities/queries/causal_insight.py:50]: That said, see file level comment.

File: cities/queries/causal_insight.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+        #                         num_samples=self.num_samples, parallel=True)
+        #                 self.samples = self.predictive(*self.model_args)
+        #     self.mwc = mwc
+
+    def generate_tensed_samples(self):
+        self.tensed_samples = {}
+        self.tensed_tau_samples = {}
+
+        for shift in [1, 2, 3]:
+            self.load_guide(shift)
```
Comment (azane) [inline cities/queries/causal_insight.py:95]: This is a bit odd to me, because e.g. `self.guide` will just be whatever you happened to load last. See file-level comment for how to restructure this.

File: cities/queries/causal_insight.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        #     self.tensed_intervened_samples[shift] = gather(
+        #     self.samples['Y'], IndexSet(**{"T": {1}}),
+        #     event_dim=0,).squeeze()#[:,self.fips_id]
+
+        #     self.tensed_outcome_difference[shift] = (
+        #     self.tensed_intervened_samples[shift] - self.tensed_observed_samples[shift]
+        #     )
+        return
+
+    def plot_predictions(
```
Comment (azane) [inline cities/queries/causal_insight.py:230]: Haven't really dug into this plotting stuff (besides looking at the plots it spits out)

File: cities/queries/causal_insight.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+            forward_shift=self.forward_shift,
+        )
+        self.model = model_cities_interaction
+
+        self.model_args = self.data["model_args"]
+
+        self.model_conditioned = pyro.condition(
+            self.model,
+            data={"T": self.data["t"], "Y": self.data["y"], "X": self.data["x"]},
+        )
```
Comment (azane) [inline cities/queries/causal_insight.py:62]: Is this necessary here? Cz there's no inference happening.

File: cities/queries/causal_insight.py
Comment (azane) [inline cities/queries/causal_insight.py:1]: I would recommend making this class less mutable. As-is, you can instantiate it, and then call various methods that fundamentally change what the query is about, and it's hard to know exactly what the intended order of execution is to initialize the object properly.

I would consider moving the execution of any methods that you only really plan on calling once into the `__init__` — setting attributes in the `__init__` and then calling a "private" method that processes those attributes accordingly e.g.`self.tensed_samples, self.tensed_tau_samples = self._generate_samples()`. `forward_shift` maybe pass in as a list of `forward_shifts` into the init? Also can be defaulted to `(1, 2, 3)`. `model` can be an argument that defaults to `None` but then is set to `model_cities_interaction` by default if no model is passed.

`get_fips_prediction` obviously needs to take its arguments, and I think you're setting them as attributes for the sake of plotting? In this case, I would make another object (or struct type thing) that is constructed in the `get_fips_prediction` method and is the target of plotting. That would better capture the 1:M relationship between this one class and the many queries it can answer (as opposed to internally hanging on to information needed to reconstruct just the most recent query)`.

Comment (azane) [reply cities/queries/causal_insight.py:1]: And for many attributes that you won't really use again, no need to hang onto them in the init, just pass them through to whatever sub-initialization method needs them. I also find it useful to make these initialization methods more functional/pure by specifying exactly which arguments they need to take. Basically as a self-documenting strategy. You've done basically this by passing `forward_shift` into `load_guide` rather than referencing `self.forward_shift`.

Comment (azane) [reply cities/queries/causal_insight.py:1]: Just to see, I pasted the class in for chatgpt and asked it what its review would be, unprompted, it's 2nd recommendation was:

```
2. Use of Class Attributes:
Avoid using too many class attributes. This class has a large number of attributes, which can make the state of the object difficult to track and understand. Try to limit class attributes to those that are essential for the class's function.
Use more local variables in methods to avoid side effects on the class state.
```

😆

File: cities/queries/causal_insight.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+        # nearest_intervention_index = min(
+        #     range(len(interventions_this_year)),
+        #     key=lambda i: abs(interventions_this_year[i] - self.intervened_value),
+        # )
+        # self.intervened_value_original = interventions_this_year_original[
+        #     nearest_intervention_index
+        # ]
+
+        self.intervened_value_percentile = round(
```
Comment (azane) [inline cities/queries/causal_insight.py:139]: The data processing stuff above can also probably be broken out into its own method.

File: cities/queries/causal_insight.py
Comment (azane) [reply cities/queries/causal_insight.py:1]: It's other suggestion was:

> 1. Separation of Concerns:
> Data Loading and Processing: It's better to separate data loading and preprocessing from the class. The class should focus on analysis, not data management.
> Model Management: Consider creating a separate class or module for handling model loading and predictions. This would make the code more modular and easier to maintain.

Whiiich I don't disagree with either.

File: cities/modeling/model_interactions.py [hunk 7]
```diff
@@ -6,71 +7,75 @@
...
-            data={"T": self.intervention, "Y": self.y, "X": self.covariates_sparse},
+            data={"T": self.data["t"], "Y": self.data["y"], "X": self.data["x"]},
+        )
+
+        self.model_rendering = pyro.render_model(
+            self.model, model_args=self.model_args, render_distributions=True
         )
 
     def train_interactions_model(self):
         self.guide = train_interactions_model(
```
Comment (azane) [inline cities/modeling/model_interactions.py:63]: As alluded to elsewhere, I think putting `self.guide = train_interactions_model(...)` in the `__init__` (and having `train_interactions_model` return the guide) might be good. Same with `self.samples` in `sample_from_guide`.

File: cities/modeling/training_pipeline.py
Comment (azane) [inline cities/modeling/training_pipeline.py:1]: This probably needs an `if __name__ == "__main__"` (to prevent it from being executed upon accidental/indirect import) and/or to be put into some folder that's explicitly set up for scripts that run as the main target (as opposed to be imported and used in a module?

File: data/model_guides/.training.log
Comment (azane) [inline data/model_guides/.training.log:1]: `.gitignore` this? Or no if want to have this info alongside the trained guides.

File: tests/test_inference.py [hunk 8]
```diff
@@ -1,18 +1,135 @@
...
-# TODO this needs to be parametrized by outcome datasets,
-# intervention datasets and forward shifts
-def test_InteractionsModel():
+# running for all is too inefficient
+# let's just run for random on a regular basis
+# comment out for a full battery of tests
+# intervention = [random.choice(interventions)]
+# outcome = [random.choice(outcomes)]
+
+interventions = [random.choice(interventions)]
```
Comment (azane) [inline tests/test_inference.py:33]: I'm not sure what the pytest-standard implementation for this is, but consider seeding this randomness so that tests consistently fail/succeed.

File: tests/test_inference.py [hunk 9]
```diff
@@ -1,18 +1,135 @@
...
+outcomes = [random.choice(outcomes)]
+
+
+# forward shifts caused some shape troubles,
+# so we test for all of them at each random test
+@pytest.mark.parametrize("intervention", interventions)
+@pytest.mark.parametrize("outcome", outcomes)
+@pytest.mark.parametrize("shift", shifts)
+def test_InteractionsModel(intervention, outcome, shift):
```
Comment (azane) [inline tests/test_inference.py:42]: Maybe call this `test_smoke_interactionsmodel` so it's more clear that it's not actually testing correctness, just that it runs.

Comment (azane) [reply tests/test_inference.py:42]: This comment applies to most of the below as well.

File: docs/guides/causal_insights_demo.ipynb
Comment (azane) [inline docs/guides/causal_insights_demo.ipynb:1]: More or less just skimmed this, but a couple comments:

1. I think most of the imports at the top of the notebook are unused and could be cleaned up.
2. Plots could do with some more axis labeling (I guess that's more of a comment for the plotting code that this notebook calls).
3. A high level explanation for the weird groupings seen in the `predictive_check` plot.

Comment (rfl-urbaniak) [reply docs/guides/causal_insights_demo.ipynb:1]: ad 1. Done with ease as now I've also added automated notebook formatting and lining. 
ad 2. done
ad 3 it's coloring by year, explanation added.

File: docs/guides/similarity_demo.ipynb
[the remark this answers is not in the document]
Comment (rfl-urbaniak) [reply docs/guides/similarity_demo.ipynb:1]: As for labelling, 

> The label on the plot "Location recent percentile" is confusing to me. Maybe switch "Location" to, e.g. Adams PA, and specify what recent means (current year?)

I can imagine some locations having longish names and messing up the graph.  What's your preference @riadas ?

[the remark this answers is not in the document]
Comment (rfl-urbaniak) [reply docs/guides/similarity_demo.ipynb:1]: @emackev re: consolidating the weights explanation, my idea was not to expose the user to everything at once, hence a gradual introduction.  I you think it's inefficient, I'll revise it, let me know.

[the remark this answers is not in the document]
Comment (rfl-urbaniak) [reply docs/guides/similarity_demo.ipynb:1]: @emackev but to alleviate some of this, I added some clarifications to the notebook, emphasizing when a functionality is not used in the front end and pointing to the weight decay parameter when it is first used.

File: cities/queries/causal_insight.py (hunk 1 above)
Comment (rfl-urbaniak) [reply cities/queries/causal_insight.py:267]: Thanks for the sanity check here! As for running full predictive, the posterior is so wide that it's not hard to sit within. :)

File: cities/queries/causal_insight.py (hunk 2 above)
Comment (rfl-urbaniak) [reply cities/queries/causal_insight.py:70]: Generally, I get the idea that attributes should be introduced upfront. However, here, the more involved attributes often require some computations and aren't always needed for all uses of an object so producing them upfront would be wasteful.  This is important especially since we care about speed and memory on the front end side.

File: cities/queries/causal_insight.py (hunk 3 above)
Comment (rfl-urbaniak) [reply cities/queries/causal_insight.py:116]: yeh, but guides are not used after the predictions are generated, so I don't really care which guide is left hanging. Will take a look at the detailed comment though.

File: cities/queries/causal_insight.py (hunk 5 above)
Comment (rfl-urbaniak) [reply cities/queries/causal_insight.py]: good catch, this was a redundant piece that was used in an earlier version. Removed.

File: cities/queries/causal_insight.py
Comment (rfl-urbaniak) [reply cities/queries/causal_insight.py:1]: I think you're right in principle about the class having grown a bit large, true that. For sure, it needs documentation as to what does what and when. But the load-or-compute-only-when-needed and not at init, and the overwrite-attributes-when-you-run-a-new-query is specifically designed to limit compute and memory use for the frontend. I'd be happy to chat about this in person, but I did consider those generic pieces of advice and intentionally went against them. The suggestion to move plotting to another class is interesting, not sure if it's worth messing with whatever Ria has already done on the frontend though.

File: cities/queries/causal_insight.py (hunk 6 above)
Comment (rfl-urbaniak) [reply cities/queries/causal_insight.py:201]: okay, I think part of why the class is kinda annoying is that I could move some of the code to external modules and call them after importing here. Will make smalls steps in this direction, for now made an issue about it.

File: cities/modeling/model_interactions.py (hunk 7 above)
Comment (rfl-urbaniak) [reply cities/modeling/model_interactions.py:64]: Here moving stuff to init would be less annoying for Ria as this is not used within the front end. But again, I did consider what you're suggesting. However, not doing every compute-heavy stuff at init gives the user (myself, say) better control of what step we're at, which is useful for instance for speed profiling, logging, etc, so I prefer to be able to call these separately.

File: cities/modeling/modeling_utils.py
[the remark this answers is not in the document]
Comment (rfl-urbaniak) [reply cities/modeling/modeling_utils.py:1]: added a docstring, in principle will keep adding docstrings when the demo day passes and this lands in main.

File: cities/modeling/training_pipeline.py
Comment (rfl-urbaniak) [reply cities/modeling/training_pipeline.py:1]: interesting. Added 

`if __name__ != "__main__":
    sys.exit()`
    
  as this doesn't require indenting chunks that already contain a bunch of indentation

File: data/model_guides/.training.log
Comment (rfl-urbaniak) [reply data/model_guides/.training.log:1]: I work on multiple machines, want to be able to access info about the last training session from any of them.

File: tests/test_inference.py (hunk 8 above)
Comment (rfl-urbaniak) [reply tests/test_inference.py:33]: I have considered this and decided against it. The stability-optimal version would be to run this for all combinations. A compute-minimal way would be to fix one particular setting, which setting the seed would effectively do. I want a middle path, where randomness goes through various combinations every now and then, so that if some particular setting is a source of a problem there is a non-negligible chance of it being eventually discovered. Identifying which combination is a problem is not a problem as it is enough to look at the log/output.

File: tests/test_inference.py (hunk 9 above)
Comment (rfl-urbaniak) [reply tests/test_inference.py]: done

File: tests/test_notebooks.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import glob
+import os
+import subprocess
+
+from cities.utils.cleaning_utils import find_repo_root
+
+root = find_repo_root()
+
+
+def test_notebook_execution():
```
[the remark this answers is not in the document]
Comment (rfl-urbaniak) [reply tests/test_notebooks.py]: it's even better now, moved out to a script.

File: cities/queries/causal_insight.py (hunk 1 above)
Comment (azane) [reply cities/queries/causal_insight.py:262]: Ah I see. I guess that's true statistically, but it would still catch serious bugs (though I think we'd have seen their effect elsewhere). Anyway — could still be useful down the road.

File: cities/queries/causal_insight.py (hunk 2 above)
Comment (azane) [reply cities/queries/causal_insight.py:70]: I would say then just set the attributes to None in the `__init__` as a form of documentation.

File: cities/queries/causal_insight.py
Comment (azane) [reply cities/queries/causal_insight.py:1]: Ya let's talk more about it. You can achieve a lot of the same design goals (like reducing attributes and side effects on class state) and still give the user control over when compute happens. If that's the goal though, a functional approach might be better and let the user track the state explicitly themselves.

File: cities/queries/causal_insight.py (hunk 3 above)
Comment (azane) [reply cities/queries/causal_insight.py:111]: This one I think would be most robustly handled by explicitly handling the currently implicit 1:M that's going on here.

> get_fips_prediction obviously needs to take its arguments, and I think you're setting them as attributes for the sake of plotting? In this case, I would make another object (or struct type thing) that is constructed in the get_fips_prediction method and is the target of plotting. That would better capture the 1:M relationship between this one class and the many queries it can answer (as opposed to internally hanging on to information needed to reconstruct just the most recent query)`.

**Review by riadas — APPROVED** [review]
Working on the frontend, hosted temporarily at http://polis-staging.basis.ai/.
